tests/end2end/fixtures/qutewm_sub.py

```diff
@@ -0,0 +1,400 @@
+                            value_mask=ev.value_mask)
+
+    def _on_circulate_request(self, ev):
+        """Called when a CirculateRequest is intercepted."""
+        ev.window.circulate(ev.place)
+
+    def _on_unmap_notify(self, ev):
+        """Called when a window is unmapped from the screen."""
+        log.debug("window unmapped: {}".format(ev.window))
+        if ev.event == self.root and not ev.from_configure:
+            log.debug("ignoring synthetic event")
+            return
+        log.info("window closed: [{:#x}]".format(ev.window.id))
+        self._unmanage_window(ev.window)
+        if self.window_stack:
+            self.activate(self.window_stack[-1])
+
+    def _on_client_message(self, ev):
+        """Called when a ClientMessage is received."""
+        if ev.client_type == self.atoms.active_window:
+            log.debug("external request to activate {}".format(ev.window))
+            if ev.window in self.windows:
+                self.activate(ev.window)
+            else:
+                log.debug("window to activate is not managed!")
+        elif ev.client_type == self.atoms.wm_state:
+            self._handle_wm_state(ev)
+
+    def _on_property_notify(self, ev):
+        """Called when a PropertyNotify event is received."""
+        if ev.atom == self.atoms.wm_hints:
+            hints = ev.window.get_wm_hints()
+            if hints is None:
+                return
+            if hints.flags & Xutil.UrgencyHint:
+                log.debug("urgency switch to {} (via WM_HINTS)"
+                          .format(ev.window))
+                self.activate(ev.window)
+
+    def _on_focus_in(self, ev):
+        """Called when a FocusIn event is received."""
+        window = ev.window
+        self.root.change_property(
+            self.atoms.active_window,
+            Xatom.WINDOW,
+            32,
+            [window.id] if window else [X.NONE],
+        )
+        self._needs_update = True
+        self._update_clients()
+        # re-order window_stack so that the active window is at
+        # window_stack[-1]
+        try:
+            index = self.window_stack.index(window)
+        except ValueError:
+            # Okay, fine then
+            pass
+        else:
+            self.window_stack = (self.window_stack[index + 1:] +
+                                 self.window_stack[:index + 1])
+        log.info("window activated: [{:#x}]".format(window.id))
+
+    def _handle_wm_state(self, ev):
+        """Handle the _NET_WM_STATE client message."""
+        client_properties = ev.window.get_property(self.atoms.wm_state,
+                                                   Xatom.ATOM, 0, 32)
+        if client_properties is None:
+            client_properties = set()
+        else:
+            client_properties = set(client_properties.value)
+
+        action = ev.data[1][0]
+        updates = {ev.data[1][1]}
+        if ev.data[1][2] != 0:
+            updates.add(ev.data[1][2])
```

Comment:
The zeroeth index is the action (add, remove, toggle), the first index is the property, and the second may contain another property or just be empty. Again, I will add a comment and a link to the relevant [spec section](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472615568):

```
  window  = the respective client window
  message_type = _NET_WM_STATE
  format = 32
  data.l[0] = the action, as listed below
  data.l[1] = first property to alter
  data.l[2] = second property to alter
  data.l[3] = source indication
  other data.l[] elements = 0
```